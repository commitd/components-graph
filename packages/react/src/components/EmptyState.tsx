import { Caption, Column, Heading, Row } from '@committed/ds'
import React from 'react'

interface EmptyStateProps {
  message?: React.ReactNode
  help?: string
  size?: 'small' | 'default'
  logo?: string
}

export const EmptyState: React.FC<EmptyStateProps> = ({
  message = 'No results found',
  size = 'default',
  help,
  logo,
}) => {
  return (
    <Row css={{ justifyContent: 'center', width: '100%', padding: '$5' }}>
      <Column css={{ alignItems: 'center' }}>
        {logo == null ? (
          <img
            src={logo}
            alt=""
            style={{
              filter: 'grayscale(75%) opacity(75%)',
              width: size === 'small' ? 100 : undefined,
            }}
          />
        ) : null}
        <Heading variant="h5" size={'$1'} css={{ mt: '$3' }}>
          {message}
        </Heading>
        {help == null ? <Caption css={{ mt: '$2' }}>{help}</Caption> : null}
      </Column>
    </Row>
  )
}
