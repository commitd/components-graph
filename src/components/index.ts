export * from './Graph'
export * from './GraphDebugControl'
export * from './GraphToolbar'
export * from './NodeViewer'
