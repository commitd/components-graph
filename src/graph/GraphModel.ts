import { ContentModel } from './ContentModel'
import { DecoratorModel } from './DecoratorModel'
import { LayoutModel } from './LayoutModel'
import { SelectionModel } from './SelectionModel'
import {
  DecoratedEdge,
  DecoratedNode,
  GraphCommand,
  ModelEdge,
  ModelNode,
} from './types'

/** The GraphModel is a declarative, immutable definition of graph state
 
The GraphModel defines:
- Nodes/edges (ContentModel)
	- No requirement for ontology, typing etc
	- Attributes
- Decoration (DecoratorModel)
	- May define custom node/edge Decorators based on any property on the node/edge
- Selection (SelectionModel)
	- Multi-select
- Graph layout (LayoutModel)
	- Force Layout / Cola JS
	- Circle
	- Grid
	- Custom (supplying a function)
- CommandQueue
	- Trigger behaviour from outside the GraphModel/React component
	- Current commands: zoom in/out, relayout, refit
 */
export class GraphModel {
  private readonly contentModel: ContentModel
  private readonly selectionModel: SelectionModel
  private readonly decoratorModel: DecoratorModel
  private readonly layoutModel: LayoutModel
  private readonly commandQueue: GraphCommand[]

  constructor(
    contentModel: ContentModel,
    options: {
      selectionModel?: SelectionModel
      decoratorModel?: DecoratorModel
      layoutModel?: LayoutModel
      commandQueue?: GraphCommand[]
    } = {}
  ) {
    this.contentModel = contentModel
    this.decoratorModel =
      options.decoratorModel ?? DecoratorModel.createDefault()
    this.layoutModel = options.layoutModel ?? LayoutModel.createDefault()
    this.selectionModel =
      options.selectionModel ?? SelectionModel.createDefault()
    this.commandQueue = options.commandQueue ?? []
  }

  static applyContent(
    graphModel: GraphModel,
    contentModel: ContentModel
  ): GraphModel {
    return new GraphModel(contentModel, {
      ...graphModel.getOptions(),
      selectionModel: GraphModel.reconcileSelection(
        contentModel,
        graphModel.getSelection()
      ),
    })
  }

  static applySelection(
    graphModel: GraphModel,
    selectionModel: SelectionModel
  ): GraphModel {
    return new GraphModel(graphModel.getCurrentContent(), {
      ...graphModel.getOptions(),
      selectionModel: GraphModel.reconcileSelection(
        graphModel.getCurrentContent(),
        selectionModel
      ),
    })
  }

  static applyLayout(
    graphModel: GraphModel,
    layoutModel: LayoutModel
  ): GraphModel {
    return new GraphModel(graphModel.contentModel, {
      ...graphModel.getOptions(),
      layoutModel,
      selectionModel: graphModel.getSelection(),
    })
  }

  static createEmpty(): GraphModel {
    return new GraphModel(ContentModel.createEmpty())
  }

  static createWithContent(contentModel: ContentModel): GraphModel {
    return new GraphModel(contentModel)
  }

  getCurrentContent(): ContentModel {
    return this.contentModel
  }

  getDecorators(): DecoratorModel {
    return this.decoratorModel
  }

  // get the current fully decorated nodes
  get nodes(): DecoratedNode[] {
    return this.decoratorModel.getDecoratedNodes(
      Object.values(this.contentModel.nodes)
    )
  }

  // get the current fully decorated edges
  get edges(): DecoratedEdge[] {
    return this.decoratorModel.getDecoratedEdges(
      Object.values(this.contentModel.edges)
    )
  }

  get selectedNodes(): DecoratedNode[] {
    return this.decoratorModel.getDecoratedNodes(
      Array.from(this.getSelection().nodes)
        .map((n) => this.contentModel.getNode(n))
        .filter((n) => n != null)
        .map((n) => n as ModelNode)
    )
  }

  get selectedEdges(): DecoratedEdge[] {
    return this.decoratorModel.getDecoratedEdges(
      Array.from(this.getSelection().edges)
        .map((e) => this.contentModel.getEdge(e))
        .filter((e) => e != null)
        .map((e) => e as ModelEdge)
    )
  }

  getNode(id: string): DecoratedNode | undefined {
    const node = this.contentModel.getNode(id)
    if (node == null) {
      return node
    }
    return this.decoratorModel.getDecoratedNodes([node])[0]
  }

  getEdge(id: string): DecoratedEdge | undefined {
    const edge = this.contentModel.getEdge(id)
    if (edge == null) {
      return edge
    }
    return this.decoratorModel.getDecoratedEdges([edge])[0]
  }

  getCurrentLayout(): LayoutModel {
    return this.layoutModel
  }

  getSelection(): SelectionModel {
    return this.selectionModel
  }

  getCommands(): GraphCommand[] {
    return this.commandQueue
  }

  pushCommand(command: GraphCommand): GraphModel {
    return new GraphModel(this.contentModel, {
      ...this.getOptions(),
      commandQueue: [...this.commandQueue, command],
    })
  }

  clearCommands(): GraphModel {
    if (this.commandQueue.length === 0) {
      return this
    }
    return new GraphModel(this.contentModel, {
      ...this.getOptions(),
      commandQueue: [],
    })
  }

  private getOptions(): {
    selectionModel: SelectionModel
    decoratorModel: DecoratorModel
    layoutModel: LayoutModel
    commandQueue: GraphCommand[]
  } {
    return {
      selectionModel: this.selectionModel,
      decoratorModel: this.decoratorModel,
      layoutModel: this.layoutModel,
      commandQueue: this.commandQueue,
    }
  }

  private static reconcileSelection(
    contentModel: ContentModel,
    selectionModel: SelectionModel
  ): SelectionModel {
    const nodes = Array.from(selectionModel.nodes).filter((n) =>
      contentModel.containsNode(n)
    )
    const edges = Array.from(selectionModel.edges).filter((n) =>
      contentModel.containsEdge(n)
    )
    return new SelectionModel(nodes, edges)
  }
}
