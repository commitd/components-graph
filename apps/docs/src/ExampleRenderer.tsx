import { Box, Button, Heading, Monospace, Row } from '@committed/components'
import { Generator, GraphRenderer, GraphRendererOptions } from '@committed/components-graph'
import React from 'react'

const Renderer: GraphRenderer<GraphRendererOptions>['render'] = ({
  graphModel,
  onChange,
}) => {
  const handleAddNode = (): void => {
    onChange(Generator.addRandomNode(graphModel))
  }

  const handleAddEdge = (): void => {
    onChange(Generator.addRandomEdge(graphModel))
  }

  return (
    <Box css={{ backgroundColor: '$paper', p: '$3' }}>
      <Button css={{ m: '$3' }} onClick={handleAddNode}>
        Add Node
      </Button>
      <Button css={{ m: '$3' }} onClick={handleAddEdge}>
        Add Edge
      </Button>
      <Row css={{ flexWrap: 'wrap', justifyContent: 'space-evenly' }}>
        <div>
          <Heading variant="h5" size={1}>
            Nodes:
          </Heading>
          <Monospace size={-2}>
            {JSON.stringify(graphModel.nodes, null, 2)}
          </Monospace>
        </div>
        <div>
          <Heading variant="h5" size={1}>
            Edges:{' '}
          </Heading>
          <Monospace size={-2}>
            {JSON.stringify(graphModel.edges, null, 2)}
          </Monospace>
        </div>
      </Row>
    </Box>
  )
}

export const exampleRenderer: GraphRenderer<GraphRendererOptions> = {
  defaultOptions: {
    height: 300,
  },
  layouts: [],
  render: Renderer,
}
