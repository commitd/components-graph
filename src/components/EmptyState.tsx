import { Flex, Heading, Text } from '@committed/components'
import React from 'react'

interface EmptyStateProps {
  message?: React.ReactNode
  help?: string
  size?: 'small' | 'default'
  logo?: string
}

export const EmptyState: React.FC<EmptyStateProps> = ({
  message = 'No results found',
  size = 'default',
  help,
  logo,
}) => {
  return (
    <Flex justifyContent="center" width={1} padding={5}>
      <Flex flexDirection="column" alignItems="center">
        {logo == null ? (
          <img
            src={logo}
            alt=""
            style={{
              filter: 'grayscale(75%) opacity(75%)',
              width: size === 'small' ? 100 : undefined,
            }}
          />
        ) : null}
        <Heading.h4 mt={3}>{message}</Heading.h4>
        {help == null ? (
          <Text variant="caption" mt={2}>
            {help}
          </Text>
        ) : null}
      </Flex>
    </Flex>
  )
}
