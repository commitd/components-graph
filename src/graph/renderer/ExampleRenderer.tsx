import { Box, Button, Heading, Monospace } from '@committed/components'
import React from 'react'
import { GraphModel } from '../GraphModel'
import { GraphRenderer, GraphRendererOptions } from '../types'

const Renderer: GraphRenderer<GraphRendererOptions>['render'] = ({
  graphModel,
  onChange,
}) => {
  const handleAddNode = (): void => {
    const withNode = graphModel.getCurrentContent().addNode({})
    const newModel = GraphModel.applyContent(graphModel, withNode)
    onChange(newModel)
  }

  const handleAddEdge = (): void => {
    const nodes = Object.values(graphModel.getCurrentContent().nodes)
    if (nodes.length === 0) {
      return
    }
    const node1 = nodes[Math.floor(Math.random() * nodes.length)]
    const node2 = nodes[Math.floor(Math.random() * nodes.length)]
    const withEdge = graphModel.getCurrentContent().addEdge({
      source: node1.id,
      target: node2.id,
    })
    const newModel = GraphModel.applyContent(graphModel, withEdge)
    onChange(newModel)
  }

  return (
    <Box bgcolor="background.paper" p={3}>
      <Button m={2} color="primary" onClick={handleAddNode}>
        Add Node
      </Button>
      <Button m={2} color="primary" onClick={handleAddEdge}>
        Add Edge
      </Button>
      <div>
        <Heading.h2>Nodes</Heading.h2>
        <Monospace fontSize={-2}>
          {JSON.stringify(graphModel.nodes, null, 2)}
        </Monospace>
      </div>
      <div>
        <Heading.h2>Edges</Heading.h2>
        <Monospace fontSize={-2}>
          {JSON.stringify(graphModel.edges, null, 2)}
        </Monospace>
      </div>
      <div></div>
    </Box>
  )
}

export const exampleRenderer: GraphRenderer<GraphRendererOptions> = {
  defaultOptions: {
    height: 300,
  },
  layouts: [],
  render: Renderer,
}
