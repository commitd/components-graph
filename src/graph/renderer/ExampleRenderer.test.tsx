import React from 'react'
import { Graph } from '../../components'
import {
  exampleGraph,
  renderDark,
  renderLight,
  userEvent,
} from '../../setupTests'
import { GraphModel } from '../GraphModel'
import { exampleRenderer } from './ExampleRenderer'

let graphModel: GraphModel

beforeEach(() => {
  graphModel = exampleGraph
})

it('can be rendered light', () => {
  const onChange = jest.fn()

  const { asFragment } = renderLight(
    <Graph
      model={graphModel}
      onModelChange={onChange}
      renderer={exampleRenderer}
      options={{ height: 'full-height' }}
    />
  )
  expect(asFragment()).toMatchSnapshot()
})

it('can be rendered light', () => {
  const onChange = jest.fn()

  const { asFragment } = renderDark(
    <Graph
      model={graphModel}
      onModelChange={onChange}
      renderer={exampleRenderer}
      options={{ height: 'full-height' }}
    />
  )
  expect(asFragment()).toMatchSnapshot()
})

it('can add node', () => {
  const onChange = jest.fn()

  const { getByText } = renderLight(
    <Graph
      model={graphModel}
      onModelChange={onChange}
      renderer={exampleRenderer}
      options={{ height: 'full-height' }}
    />
  )

  userEvent.click(getByText('Add Node'))
  expect(onChange).toHaveBeenCalled()
})

it('can add Edge', () => {
  const onChange = jest.fn()

  const { getByText } = renderLight(
    <Graph
      model={graphModel}
      onModelChange={onChange}
      renderer={exampleRenderer}
      options={{ height: 'full-height' }}
    />
  )

  userEvent.click(getByText('Add Edge'))
  expect(onChange).toHaveBeenCalled()
})

it('can not add edge if empty', () => {
  const onChange = jest.fn()
  graphModel = GraphModel.createEmpty()

  const { getByText } = renderLight(
    <Graph
      model={graphModel}
      onModelChange={onChange}
      renderer={exampleRenderer}
      options={{ height: 'full-height' }}
    />
  )

  userEvent.click(getByText('Add Edge'))
  expect(onChange).toHaveBeenCalledWith(graphModel)
})
