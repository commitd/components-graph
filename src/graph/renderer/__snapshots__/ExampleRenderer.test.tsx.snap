// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can be rendered light 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root MuiBox-root Styled(MuiBox)-root Styled(MuiBox)-root"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained ForwardRef(withPositioning)-root ForwardRef(withPositioning)-root MuiButton-containedPrimary"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Add Node
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained ForwardRef(withPositioning)-root ForwardRef(withPositioning)-root MuiButton-containedPrimary"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Add Edge
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <div>
      <h2
        class="MuiTypography-root ForwardRef(withPositioning)-root ForwardRef(withPositioning)-root Component-root Component-root MuiTypography-h2"
      >
        Nodes
      </h2>
      <pre
        class="ForwardRef(withPositioning)-root ForwardRef(withPositioning)-root Component-root Component-root"
        variant="body1"
      >
        [
  {
    "id": "node1",
    "attributes": {
      "employer": "Committed"
    },
    "color": "yellow",
    "label": "Node 1",
    "size": 10,
    "strokeColor": "black",
    "opacity": 1,
    "shape": "ellipse",
    "strokeSize": 2
  },
  {
    "id": "node2",
    "attributes": {
      "employer": "Government"
    },
    "color": "green",
    "label": "Node 2",
    "size": 12,
    "strokeColor": "black",
    "opacity": 0.9,
    "shape": "rectangle",
    "strokeSize": 3
  }
]
      </pre>
    </div>
    <div>
      <h2
        class="MuiTypography-root ForwardRef(withPositioning)-root ForwardRef(withPositioning)-root Component-root Component-root MuiTypography-h2"
      >
        Edges
      </h2>
      <pre
        class="ForwardRef(withPositioning)-root ForwardRef(withPositioning)-root Component-root Component-root"
        variant="body1"
      >
        [
  {
    "id": "edge1",
    "attributes": {
      "role": "client"
    },
    "source": "node1",
    "target": "node2"
  }
]
      </pre>
    </div>
    <div />
  </div>
</DocumentFragment>
`;

exports[`can be rendered light 2`] = `
<DocumentFragment>
  <div
    class="MuiBox-root MuiBox-root Styled(MuiBox)-root Styled(MuiBox)-root"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained ForwardRef(withPositioning)-root ForwardRef(withPositioning)-root MuiButton-containedPrimary"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Add Node
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained ForwardRef(withPositioning)-root ForwardRef(withPositioning)-root MuiButton-containedPrimary"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Add Edge
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <div>
      <h2
        class="MuiTypography-root ForwardRef(withPositioning)-root ForwardRef(withPositioning)-root Component-root Component-root MuiTypography-h2"
      >
        Nodes
      </h2>
      <pre
        class="ForwardRef(withPositioning)-root ForwardRef(withPositioning)-root Component-root Component-root"
        variant="body1"
      >
        [
  {
    "id": "node1",
    "attributes": {
      "employer": "Committed"
    },
    "color": "yellow",
    "label": "Node 1",
    "size": 10,
    "strokeColor": "black",
    "opacity": 1,
    "shape": "ellipse",
    "strokeSize": 2
  },
  {
    "id": "node2",
    "attributes": {
      "employer": "Government"
    },
    "color": "green",
    "label": "Node 2",
    "size": 12,
    "strokeColor": "black",
    "opacity": 0.9,
    "shape": "rectangle",
    "strokeSize": 3
  }
]
      </pre>
    </div>
    <div>
      <h2
        class="MuiTypography-root ForwardRef(withPositioning)-root ForwardRef(withPositioning)-root Component-root Component-root MuiTypography-h2"
      >
        Edges
      </h2>
      <pre
        class="ForwardRef(withPositioning)-root ForwardRef(withPositioning)-root Component-root Component-root"
        variant="body1"
      >
        [
  {
    "id": "edge1",
    "attributes": {
      "role": "client"
    },
    "source": "node1",
    "target": "node2"
  }
]
      </pre>
    </div>
    <div />
  </div>
</DocumentFragment>
`;
